### PR TITLE
Fix offset calculation in VarBinaryEncoder for slices

### LINKED ISSUE #126: Crashes over "read over past end of file"
```
Traceback (most recent call last):
  File "/home/ubuntu/work/lance/python/benchmarks/./oxford_pet.py", line 105, in <module>
    main()
  File "/home/ubuntu/work/lance/python/venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/ubuntu/work/lance/python/venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/ubuntu/work/lance/python/venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/ubuntu/work/lance/python/venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/ubuntu/work/lance/python/benchmarks/bench_utils.py", line 158, in main
    run_benchmark(b)
  File "/home/ubuntu/work/lance/python/benchmarks/bench_utils.py", line 151, in run_benchmark
    b.run(base_uri=base_uri, fmt=f, flavor=flavor)
  File "/home/ubuntu/work/lance/python/benchmarks/bench_utils.py", line 180, in run
    output = func(*args, **kwargs)
  File "/home/ubuntu/work/lance/python/benchmarks/bench_utils.py", line 191, in timeit_wrapper
    result = func(*args, **kwargs)
  File "/home/ubuntu/work/lance/python/benchmarks/./oxford_pet.py", line 30, in total_count
    return ds.count_rows()
  File "pyarrow/_dataset.pyx", line 380, in pyarrow._dataset.Dataset.count_rows
  File "pyarrow/_dataset.pyx", line 2636, in pyarrow._dataset.Scanner.count_rows
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 115, in pyarrow.lib.check_status
OSError: Cannot read past end of file
```

Running `SELECT COUNT(1) FROM ds` over the oxford pet dataset.

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #126

### CODE REVIEW

File: cpp/src/lance/encodings/binary.cc [hunk 1]
```diff
@@ -41,12 +41,18 @@ Result<int64_t> VarBinaryEncoder::Write(const std::shared_ptr<::arrow::Array> da
   ARROW_ASSIGN_OR_RAISE(auto offsets_position, out_->Tell());
   offsets_builder_.Reset();
   assert(arr->length() > 0);
+  /// Reset the slice's first offset to zero.
+  auto first_offset = arr->value_offset(0);
   for (int64_t i = 0; i <= arr->length(); ++i) {
-    ARROW_RETURN_NOT_OK(offsets_builder_.Append(start_offset + arr->value_offset(i)));
+    ARROW_RETURN_NOT_OK(
+        offsets_builder_.Append(start_offset + arr->value_offset(i) - first_offset));
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/binary.cc:48]: What's start offset vs first offset?

File: cpp/src/lance/encodings/binary.cc [hunk 2]
```diff
@@ -41,12 +41,18 @@ Result<int64_t> VarBinaryEncoder::Write(const std::shared_ptr<::arrow::Array> da
...
   }
   ARROW_RETURN_NOT_OK(offsets_builder_.Finish(&offsets_));
+  assert(offsets_->Value(0) == start_offset);
 
-  ARROW_RETURN_NOT_OK(out_->Write(offsets_->values()));
+  ARROW_RETURN_NOT_OK(
+      out_->Write(offsets_->raw_values(),
+                  (arr->length() + 1) * sizeof(typename ::arrow::TypeTraits<OffsetType>::CType)));
```
Comment (changhiskhan) [inline cpp/src/lance/encodings/binary.cc:55]: What does this mean?

**Review by changhiskhan — COMMENTED** [review]
2 minor questions for my own edification

File: cpp/src/lance/encodings/binary.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/binary.cc:50]: start_offset is the offset in the output file

first_offset is the "offset" of the value in the "in-memory array".

I can change these terms if confusing

File: cpp/src/lance/encodings/binary.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/encodings/binary.cc:57]: `offsets_->values()` is the underneath buffer shared by potentially multiple arrays.

`raw_values()` is like `values() + buffer.offset` which point to the data that this exact array holds.

```
sizeof(typename ::arrow::TypeTraits<OffsetType>::CType)
```

so it basically write `(len + 1)` offsets into disk.

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**